nixos/modules/services/misc/radicle.nix

```diff
@@ -0,0 +1,344 @@
+{ config, lib, pkgs, ... }:
+with lib;
+let
+  cfg = config.services.radicle;
+
+  json = pkgs.formats.json { };
+
+  env = rec {
+    # rad fails if it cannot stat $HOME/.gitconfig
+    HOME = "/var/lib/radicle";
+    RAD_HOME = HOME;
+  };
+
+  # Convenient wrapper to run `rad` in the namespaces of `radicle-node.service`
+  rad-system = pkgs.writeShellScriptBin "rad-system" ''
+    set -o allexport
+    ${toShellVars env}
+    # Note that --env is not used to preserve host's envvars like $TERM
+    exec ${getExe' pkgs.util-linux "nsenter"} -a \
+      -t "$(${getExe' config.systemd.package "systemctl"} show -P MainPID radicle-node.service)" \
+      -S "$(${getExe' config.systemd.package "systemctl"} show -P UID radicle-node.service)" \
+      -G "$(${getExe' config.systemd.package "systemctl"} show -P GID radicle-node.service)" \
+      ${getExe' cfg.package "rad"} "$@"
+  '';
+
+  commonServiceConfig = serviceName: {
+    environment = env // {
+      RUST_LOG = mkDefault "info";
+    };
+    path = [
+      pkgs.gitMinimal
+    ];
+    documentation = [
+      "https://docs.radicle.xyz/guides/seeder"
+    ];
+    after = [
+      "network.target"
+      "network-online.target"
+    ];
+    requires = [
+      "network-online.target"
+    ];
+    wantedBy = [ "multi-user.target" ];
+    serviceConfig = mkMerge [
+      {
+        BindReadOnlyPaths = [
+          "${cfg.configFile}:${env.RAD_HOME}/config.json"
+          "${if isPath cfg.publicKeyFile then cfg.publicKeyFile else pkgs.writeText "radicle.pub" cfg.publicKeyFile}:${env.RAD_HOME}/keys/radicle.pub"
+        ];
+        KillMode = "process";
+        StateDirectory = [ "radicle" ];
+        DynamicUser = true;
+        # The "radicale" user will be allocated when the first of the "radicle-*" services starts,
+        # it will be shared among them, and be released when the last one of them exits.
```

Comment:
Especially when multiple services share the same user, using DynamicUser can cause some strange behaviors.

Reply:
@SuperSandro2000, what strange behavior do you have in mind? because it's [a setup suggested by Lennart Poettering](https://github.com/systemd/systemd/issues/7253#issuecomment-344289925):

> note that you can actually run multiple independent services as the same dynamic user. Wouldn't that already get you quite far? i.e. let's say you have five services that shall be able to mess around with each other files. Simply set DynamicUser=1 for all of them, but combine this with User= set to some common, clear user name. If you do, then the user will be allocated when the first of these services starts, and be released when the last one of them exits.